README.md
# osm-static-maps

Openstreetmap static maps is a Node.js lib, CLI and server open source inspired on Google static map service.

## How to use

### 1. CLI

```bash
sudo npm i -g osm-static-maps
osmsm --help
osmsm -g '{"type":"Point","coordinates":[-105.01621,39.57422]}' > map.png
```

* note: if you have this error trying to install globally `Error: EACCES: permission denied, mkdir '/usr/local/lib/node_modules/osm-static-maps/node_modules/puppeteer/.local-chromium'`, it's caused by this pupeteer issue https://github.com/puppeteer/puppeteer/issues/367, you can workaround by installing globally with the `unsafe-perm` flag:

```bash
sudo npm i -g osm-static-maps --unsafe-perm=true
```

### 2. Node.js library

```bash
npm install osm-static-maps
```

```javascript
// index.js old school
const { osmsm } = require('osm-static-maps');
osmsm({geojson: geojson})
  .then(function(imageBinaryBuffer) { ... })
  .catch(function(error) { ... });

// index.js modern style (also supports typescript)
import { osmsm } from 'osm-static-maps';
const imageBinaryBuffer = await osmsm({geojson});
```

### 3. Standalone sample server

```bash
sudo npm i -g osm-static-maps
osmsm serve
```

Or you can use docker-compose
```bash
git clone git@github.com:jperelli/osm-static-maps.git
cd osm-static-maps
docker-compose up
```

## API Reference

All parameters have a short and long version. The short version can be used only with the shell CLI. The long version can be used with the library and can be passed to the app server as GET query params, or POST json body (remember to set the header `Content-Type: application/json`)

|   | Parameter | Description | Default Value |
| - | ---- | ---- | ---- |
| g | geojson | geojson object to be rendered in the map | `undefined` |
| f | geojsonfile | filename or url to read geojson data from (use '-' to read from stdin on CLI) | `undefined` |
| H | height | height in pixels of the returned img | `600` |
| W | width | height in pixels of the returned img | `800` |
| c | center | center of the map lon,lat floats string | (center of the geojson) or `'-57.9524339,-34.921779'` |
| z | zoom | zoomlevel of the leaflet map | if `vectorserverUrl` available, use `12` else `20` |
| Z | maxZoom | max zoomlevel of the leaflet map | `17` |
| A | attribution | attribution legend | `'osm-static-maps / © OpenStreetMap contributors'` |
| t | tileserverUrl | url of a tileserver | `'http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png'` |
| m | vectorserverUrl | url of a vector tile server (MVT style.json) | `undefined` |
| M | vectorserverToken | token of the vector tile server (MVT) | `'no-token'` |
| D | renderToHtml | returns html of the webpage containing the map (instead of a binary image) | `false` |
| F | type | format of the image returned (`'jpeg'`/`'png'`) | `'png'` |
| q | quality | quality of the image returned (`0`-`100`, only for `jpg`) | `100` |
| x | imagemin | enable lossless compression with [optipng](https://github.com/imagemin/imagemin-optipng) / [jpegtran](https://github.com/imagemin/imagemin-jpegtran) | `false` |
| a | arrows | render arrows to show the direction of linestrings | `false` |
| s | scale | enable render a scale ruler (boolean or [a json options object](https://leafletjs.com/reference-1.6.0.html#control-scale-option)) | `false` |
| T | timeout | miliseconds until page load throws timeout | `20000` |
| k | markerIconOptions | set marker icon options ([a json options object](https://leafletjs.com/reference-1.6.0.html#icon-option)) *see note | `undefined` (leaflet's default marker) |
| S | style | style to apply to each feature ([a json options object](https://leafletjs.com/reference-1.6.0.html#path-option)) *see note | `undefined` (leaflet's default) |
| e | haltOnConsoleError | throw error if there is any `console.error(...)` when rendering the map image | `false` |

* Note on markerIconOptions: it's also accepted a markerIconOptions attribute in the geojson feature, for example
```json
{"type":"Point","coordinates":[-105.01621,39.57422],"markerIconOptions":{"iconUrl":"https://leafletjs.com/examples/custom-icons/leaf-red.png"}}
```

* Note on style: it's also accepted a pathOptions attribute in the geojson feature, for example
```json
{"type":"Polygon","coordinates":[[[-56.698,-36.413],[-56.716,-36.348],[-56.739,-36.311]]],"pathOptions":{"color":"#FF5555"}}
```

* also remember that the `#` char needs to be passed as `%23` if you are using GET params

## Design considerations & architecture

[Read the blogpost](https://jperelli.com.ar/project/2019/10/01/osm-static-maps/) on the creation of this library and how it works internally.

## LICENSE

 - GPLv2

## Credits

Specially to the contributors of

- OpenStreetMap
- Leaflet
- Puppeteer
- ExpressJS
- Handlebars


CHANGELOG.md
# 3.11.0

- updated dependencies
- no default export

# 3.10.3

Support AWS functions hotfix

 - Move dependencies right.

# 3.10.2

Support AWS functions

 - Move dependencies to dev to conform the 50Mib limit

# 3.10.1

Hotfix

 - Fix broken local puppeteer

# 3.10.0

Add support to vercel api functions

 - Use chrome-aws-lambda as alternative to base puppeteer

# 3.9.1

Maintenance

 - Upgrade dependencies to latest versions

# 3.9.0

Read geojson from file or stdin

 - Add new option `-f` to read geojson from file, stdin (`-`) or url
 - Fix issue on console error misses, add `haltOnConsoleError` option
 - Prevent HTML/JS injection vulnerability in sample server
 - Upgrade dependencies

# 3.8.1

Maintenance

 - Upgrade dependencies, outstandingly puppeteer from 3.x to 5.x
 - Improve sample-server logging client information

# 3.8.0

Add styling options to features

 - Add per feature and global style configuration option
 - Add timeout option
 - Add Dockerfile kanji font support
 - Add geojson typescript type
 - Fix arrows argument crashing lib
 - Update all dependencies to the latest version

# 3.7.0

Add serve command to CLI

 - Add serve subcommand to CLI
 - Fix sample-server
 - Updated all dependencies to latest version

# 3.6.0

New CLI interface, puppeteer singleton

 - Add CLI interface
 - Fix sample server using too many resources, make headless browser singleton
 - Updated all dependencies to latest version
 - Refactored README
 - Refactored project directory structure

# 3.5.1

Fix geojson option

 - Fix geojson option being an object was not being parsed right and throwed an error

# 3.5.0

Add scale and marker icon options

 - Add scale parameter to render a scale ruler
 - Add marker icon options

# 3.4.1

Improve code reliability

 - Improved requirements' path resolution: use builtin require() function instead of relying in custom paths
 - Updated all dependencies to latest version
 - Fix center parameter issue in sample server

# 3.4.0

Add typescript definitions, allow POST in server

 - Add typescript definitions
 - Allow sample-server to use POST json enconded body to get the parameters
 - Allow library to be called without any arguments

# 3.3.0

Add arrows option on linestrings

 - Add `arrows` boolean option to show the direction of each linestring inside the geojson
 - Updated all dependencies to latest version
 - Improved nodemon to watch html template file
 - Remove annoying package-lock.json and pin all versions directly

# 3.2.0

Add optional lossless compresion

 - Add `imagemin` boolean option to lossless compress the result image using [imagemin](https://github.com/imagemin/imagemin) (optipng / jpegtran)
 - Add `oxipng` boolean option to lossless compress the result image using [oxipng](https://github.com/shssoichiro/oxipng) (only for png images) (searches in `/root/.cargo/bin/oxipng`, see how to install it in [Dockerfile](./Dockerfile))

# 3.1.1

Bugfixing

 - Fix bug wrong zooming when no geojson
 - Default to different zoom levels depending on vector vs raster tiles

# 3.1.0

Add mapbox vector tiles and expose more options

 - Add CHANGELOG file to track version changes
 - Add `vectorserverUrl` and `vectorserverToken` options for taking a screenshot using Mapbox Vector Tiles (MVT) Layer (style.json)
 - Expose `center` option
 - Expose `zoom` and`maxZoom` options
 - Expose `attribution` option to set the legend / copyright
 - Expose `renderToHtml` option
 - Expose `type` and `quality` to control returned image format

# 3.0.0

Upgrade packages and improve speed

 - Replace unmaintained node-webshot with puppeteer
 - Upgrade all dependencies, including leaflet
 - Fix bug geojson `Point`s were not being rendered
 - Add docker-compose support
 - Upgrade heroku config to latest CEDAR stack

# 2.1.0

Add options

 - Add `tileserverUrl` option
 - Add Dockerfile

# 2.0.0

Make it reusable as a library

- Refactored to be a npm library
- Publish package in npm
- Started using semver

# 1.0.0

Improve reliability for a stable release

- Serve leaflet locally instead of using a cdn
- Upgrade packages to latest versions
- Improve readme

# 0.0.1

Initial useful version.

- Express application that takes a screenshot of a geojson over a leafletmap with parameters `geojson`, `height`, `width`
- Published in heroku


test/index.js
const test = require('ava');
const sharp = require('sharp');
const { osmsm } = require('../src/lib');

test('static map from geojson as png', async (t) => {
  const imageBuffer = await osmsm({
    geojson: '{"type":"Point","coordinates":[-105.01621,39.57422]}',
  });

  const metadata = await sharp(imageBuffer).metadata();

  t.is(metadata.width, 800, 'image width');
  t.is(metadata.height, 600, 'image height');
  t.is(metadata.format, 'png', 'image format');
});

test('static map from center as jpeg', async (t) => {
  const imageBuffer = await osmsm({
    center: '-105.01621,39.57422',
    zoom: 12,
    width: 1280,
    height: 720,
    type: 'jpeg'
  });

  const metadata = await sharp(imageBuffer).metadata();

  t.is(metadata.width, 1280, 'image width');
  t.is(metadata.height, 720, 'image height');
  t.is(metadata.format, 'jpeg', 'image format');
});


src/cli.js
#!/usr/bin/env node

const program = require('commander');
const { osmsm } = require('./lib');
const pkg = require('../package.json');

program
  .version(pkg.version)
  .name("osmsm")
  .usage(
    "[options]\nGenerate an image of a geojson with a background map layer"
  )
  .option("-g, --geojson <string>", "Geojson object to be rendered")
  .option("-f, --geojsonfile <string>", "Geojson file name to be rendered (\"-\" reads STDIN)")
  .option(
    "-H, --height <number>",
    "Height in pixels of the returned img",
    Number,
    600
  )
  .option(
    "-W, --width <number>",
    "Width in pixels of the returned img",
    Number,
    800
  )
  .option(
    "-c, --center <lon,lat>",
    "Center of the map (default: center of geojson or '-57.9524339,-34.921779'"
  )
  .option(
    "-z, --zoom <number>",
    "Zoomlevel of the map (default: vectorserverUrl ? 12 : 20)",
    Number
  )
  .option("-Z, --maxZoom <number>", "Maximum zoomlevel of the map", Number, 17)
  .option(
    "-A, --attribution <string>",
    "Attribution legend watermark",
    "osm-static-maps / © OpenStreetMap contributors"
  )
  .option(
    "-t, --tileserverUrl <string>",
    "Url of a tileserver",
    "http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
  )
  .option(
    "-m, --vectorserverUrl <string>",
    "Url of a vector tile server (MVT style.json)"
  )
  .option(
    "-M, --vectorserverToken <string>",
    "Token of the vector tile server (MVT)"
  )
  .option(
    "-D, --renderToHtml",
    "Returns html of the webpage containing the leaflet map (instead of a binary image)",
    false
  )
  .option("-F, --type <jpeg|png>", "Format of the image returned", "png")
  .option(
    "-q, --quality <number>",
    "Quality of the image returned (only for -F=jpeg)",
    Number,
    100
  )
  .option(
    "-x, --imagemin",
    "Apply lossless compression with optipng/jpegtran",
    false
  )
  .option("-X, --oxipng", "Apply lossless compression with oxipng", false)
  .option(
    "-a, --arrows",
    "Render arrows to show the direction of linestrings",
    false
  )
  .option(
    "-s, --scale [json string]",
    "Enable render a scale ruler (see options in https://leafletjs.com/reference-1.6.0.html#control-scale-option)",
    false
  )
  .option(
    "-k, --markerIconOptions <json string>",
    "Set marker icon options (see doc in https://leafletjs.com/reference-1.6.0.html#icon-option)"
  )
  .option(
    "-T, --timeout <number>",
    "Miliseconds until page load throws timeout",
    Number,
    20000
  )
  .option(
    "-S, --style <json string>",
    "Set path style options (see doc in https://leafletjs.com/reference-1.6.0.html#path-option)"
  )
  .option(
    "-e, --haltOnConsoleError",
    "throw error if there is any console.error(...) when rendering the map image",
    false
  )
  .action(function(cmd) {
    const opts = cmd.opts();

    // DEBUG
    // process.stderr.write(JSON.stringify(opts, undefined, 2));
    // process.stderr.write("\n");

    if (!["png", "jpeg"].includes(opts.type)) {
      process.stderr.write(
        '-F|--type can only have the values "png" or "jpeg"\n'
      );
      process.exit(2);
    }

    osmsm(opts)
      .then(v => {
        process.stdout.write(v);
        process.stdout.end();
        process.exit(0);
      })
      .catch(e => {
        process.stderr.write(e.toString());
        process.exit(1);
      });
  });

program
  .command("serve")
  .option("-p, --port <number>", "Port number to listen")
  .action(function(cmd) {
    const server = require("./server");
    const port = cmd.port || process.env.PORT || 3000;
    server.listen(port, function() {
      console.log("osmsm server listening on port " + port);
    });
  });

program.on("--help", function() {
  process.stdout.write("\n");
  process.stdout.write("Examples: \n");
  process.stdout.write(
    `  $ osmsm -g '{"type":"Point","coordinates":[-105.01621,39.57422]}'\n`
  );
  process.stdout.write(
    `  $ osmsm -g '[{"type":"Feature","properties":{"party":"Republican"},"geometry":{"type":"Polygon","coordinates":[[[-104.05,48.99],[-97.22,48.98],[-96.58,45.94],[-104.03,45.94],[-104.05,48.99]]]}},{"type":"Feature","properties":{"party":"Democrat"},"geometry":{"type":"Polygon","coordinates":[[[-109.05,41.00],[-102.06,40.99],[-102.03,36.99],[-109.04,36.99],[-109.05,41.00]]]}}]' --height=300 --width=300\n`
  );
  process.stdout.write(
    `  $ osmsm -f /path/to/my_file.json\n`
  );
  process.stdout.write(
    `  $ program_with_geojson_on_stdout | osmsm -f -\n`
  );
  process.stdout.write("\n");
});

program.parse(process.argv);


src/lib.d.ts
import { Control, IconOptions, PathOptions } from 'leaflet';
import { GeoJSON } from 'geojson';

export interface OsmStaticMapsOptions {
  /**
   * Geojson object to be rendered in the map
   * @defaultValue `undefined`
   */
  geojson?: string | GeoJSON;

  /**
   * Filename or url of a geojson to be rendered in the map ('-' to use stdin)
   * @defaultValue `undefined`
   */
  geojsonfile?: string;

  /**
   * height in pixels of the returned img
   * @defaultValue `600`
   */
  height?: number;

  /**
   * width in pixels of the returned img
   * @defaultValue `800`
   */
  width?: number;

  /**
   * center of the map lon,lat floats string
   * @defaultValue (center of the geojson) or `'-57.9524339,-34.921779'`
   */
  center?: string;

  /**
   * zoomlevel of the leaflet map
   * @defaultValue `vectorserverUrl` ? `12` : `20`
   */
  zoom?: number;

  /**
   * max zoomlevel of the leaflet map
   * @defaultValue `17`
   */
  maxZoom?: number;

  /**
   * attribution legend watermark
   * @defaultValue `'osm-static-maps / © OpenStreetMap contributors'`
   */
  attribution?: string;

  /**
   * url of a tileserver
   * @defaultValue `'http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png'`
   */
  tileserverUrl?: string;

  /**
   * url of a vector tile server (MVT style.json)
   * @defaultValue `undefined`
   */
  vectorserverUrl?: string;

  /**
   * token of the vector tile server (MVT)
   * @defaultValue `'no-token'`
   */
  vectorserverToken?: string;

  /**
   * returns html of the webpage containing the map (instead of a binary image)
   * @defaultValue `false`
   */
  renderToHtml?: boolean;

  /**
   * format of the image returned
   * @defaultValue `'png'`
   */
  type?: "jpeg" | "png";

  /**
   * quality of the image returned (`0`-`100`, only for `jpg`)
   * @defaultValue `100`
   */
  quality?: number;

  /**
   * enable lossless compression with [optipng](https://github.com/imagemin/imagemin-optipng) / [jpegtran](https://github.com/imagemin/imagemin-jpegtran)
   * @defaultValue `false`
   */
  imagemin?: boolean;

  /**
   * render arrows to show the direction of linestrings
   * @defaultValue `false`
   */
  arrows?: boolean;

  /**
   * enable render a scale ruler (boolean or [a json options object](https://leafletjs.com/reference-1.6.0.html#control-scale-option))
   * @defaultValue `false`
   */
  scale?: boolean | Control.ScaleOptions;

  /**
   * set marker icon options ([a json options object](https://leafletjs.com/reference-1.6.0.html#icon-option))
   * @defaultValue `undefined`
   */
  markerIconOptions?: IconOptions;

  /**
   * miliseconds until page load throws timeout
   * @defaultValue `20000`
   */
  timeout?: number;

  /**
   *
   * @defaultValue `undefined`
   */
  style?: PathOptions;

  /**
   * throw error if there is any `console.error(...)` when rendering the map image
   *
   * @defaultValue `false`
   */
  haltOnConsoleError?: Boolean;
}

/** Renders a map controlled by the options passed and returns an image */
export declare function osmsm<T extends OsmStaticMapsOptions>(
  options?: T
): Promise<T extends { renderToHtml: true } ? string : Buffer>;


src/server.js
const express = require('express');
const http = require('http');
const { osmsm } = require('./lib.js');

const app = express();
// app.set("port", process.env.PORT || 3000);
app.set('views', __dirname + '/lib');
app.set('view engine', 'handlebars');
app.set('view options', { layout: false });
app.use(express.json({ limit: '50mb' }));

app.use((req, res, next) => {
  const ip = req.headers["x-forwarded-for"] || req.socket.remoteAddress;
  const date = new Date().toISOString();
  const ref = req.header("Referer");
  const ua = req.header("user-agent");
  const url = req.originalUrl;
  const logLine = `[${date} - ${ip}] (${ref}) {${ua}} ${url}`;
  if (process.env.HEADER_CHECK) {
    const header = process.env.HEADER_CHECK.split(":");
    if (req.headers[header[0]] !== header[1]) {
      res
        .status(403)
        .send(
          process.env.HEADER_CHECK_FAIL_MESSAGE ||
            "Forbidden, set correct header to access"
        );
      console.log(`${logLine} FORBIDDEN, HEADER_CHECK FAILED`);
      return;
    }
  }
  console.log(logLine);
  next();
});


function htmlEscape(text) {
  return text.replace(/&/g, '&amp;').
  replace(/</g, '&lt;').
  replace(/"/g, '&quot;').
  replace(/'/g, '&#039;');
}

function sanitize(params) {
  let result = {}
  for (let [key, value] of Object.entries(params)) {
      result[key] = htmlEscape(value)
  }
  return result;
}

app.get("/health", (req, res) => res.sendStatus(200));

const handler = (res, params) => {
  const filename = params.f || params.geojsonfile
  if (
    filename &&
    !(filename.startsWith("http://") ||
    filename.startsWith("https://"))
  ) {
    throw new Error(
      `'geojsonfile' parameter on server only allowed if filename starts with http(s)`
    );
  }
  osmsm(params)
    .then((data) => res.end(data))
    .catch((err) => res.status(500).end(err.toString()));
};

app.get("/", (req, res) => handler(res, req.query));
app.post("/", (req, res) => handler(res, req.body));

app.get("/dynamic", (req, res) => {
  var sanitized = sanitize(req.query)
  handler(res, { ...sanitized, renderToHtml: true })
})

app.post("/dynamic", (req, res) => {
  var sanitized = sanitize(req.body)
  handler(res, { ...sanitized, renderToHtml: true })
})

module.exports = http.createServer(app);


src/lib.js
const fs = require('fs');
const http = require('http');
const https = require('https');
const Handlebars = require('handlebars');
const path = require('path');
const puppeteer = require('puppeteer');

const files = {
  leafletjs: fs.readFileSync(require.resolve('leaflet/dist/leaflet.js'), 'utf8'),
  leafletcss: fs.readFileSync(require.resolve('leaflet/dist/leaflet.css'), 'utf8'),
  leafletpolylinedecorator: fs.readFileSync(require.resolve('leaflet-polylinedecorator/dist/leaflet.polylineDecorator.js'), 'utf8'),
  mapboxjs: fs.readFileSync(require.resolve('mapbox-gl/dist/mapbox-gl.js'), 'utf8'),
  mapboxcss: fs.readFileSync(require.resolve('mapbox-gl/dist/mapbox-gl.css'), 'utf8'),
  leafletmapboxjs: fs.readFileSync(require.resolve('mapbox-gl-leaflet/leaflet-mapbox-gl.js'), 'utf8'),
  markericonpng: new Buffer.from(fs.readFileSync(require.resolve('leaflet/dist/images/marker-icon.png')), 'binary').toString('base64'),
}
const templatestr = fs.readFileSync(path.join(__dirname, 'template.html'), 'utf8')
const template = Handlebars.compile(templatestr);

function replacefiles(str) {
  const ff = Object.entries(files)
  let res = str
  ff.reverse()
  ff.forEach(([k, v]) => res = res.replace(`//${k}//`, v))
  return res
}

function httpGet(url) {
  // from https://stackoverflow.com/a/41471647/912450
  const httpx = url.startsWith('https') ? https : http;
  return new Promise((resolve, reject) => {
    let req = httpx.get(url, (res) => {
      if (res.statusCode !== 200) {
        reject(`Error ${res.statusCode} trying to get geojson file from ${url}`);
      }
      res.setEncoding('utf8');
      let rawData = '';
      res.on('data', (chunk) => { rawData += chunk; });
      res.on('end', () => resolve(rawData) );
    });
    req.on('error', (err) => {
      reject(err);
    });
  });
}

process.on('warning', (e) => console.warn(e.stack));

async function osmsm(options) {
  options = options || {};
  options.geojson = (options.geojson && (typeof options.geojson === 'string' ? options.geojson : JSON.stringify(options.geojson))) || '';
  options.geojsonfile = options.geojsonfile || '';
  options.height = options.height || 600;
  options.width = options.width || 800;
  options.center = options.center || '';
  options.zoom = options.zoom || '';
  options.maxZoom = options.maxZoom || (options.vectorserverUrl ? 20 : 17);
  options.attribution = options.attribution || 'osm-static-maps | © OpenStreetMap contributors';
  options.tileserverUrl = options.tileserverUrl || 'http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png';
  options.vectorserverUrl = options.vectorserverUrl || '';
  options.vectorserverToken = options.vectorserverToken || 'no-token';
  options.imagemin = options.imagemin || false;
  options.oxipng = options.oxipng || false;
  options.arrows = options.arrows || false;
  options.scale = (options.scale && (typeof options.scale === 'string' ? options.scale : JSON.stringify(options.scale))) || false;
  options.markerIconOptions = (options.markerIconOptions && (typeof options.markerIconOptions === 'string' ? options.markerIconOptions : JSON.stringify(options.markerIconOptions))) || false;
  options.style = (options.style && (typeof options.style === 'string' ? options.style : JSON.stringify(options.style))) || false;
  options.timeout = typeof options.timeout === undefined ? 20000 : options.timeout;
  options.haltOnConsoleError = !!options.haltOnConsoleError;

  if (options.geojsonfile) {
    if (options.geojson) {
      throw new Error(`Only one option allowed: 'geojsonfile' or 'geojson'`)
    }
    if (options.geojsonfile.startsWith("http://") || options.geojsonfile.startsWith("https://")) {
      options.geojson = await httpGet(options.geojsonfile)
    }
    else {
      options.geojson = fs.readFileSync(
        options.geojsonfile === "-"
          ? process.stdin.fd
          : options.geojsonfile,
        "utf8"
      );
    }
  }

  const html = replacefiles(template(options));

  if (options.renderToHtml) {
    return html;
  }

  const browser = await puppeteer.launch({
    args: ['--no-sandbox', '--disable-setuid-sandbox'],
    headless: true,
    ignoreHTTPSErrors: true,
  });
  const page = await browser.newPage();

  try {
    page.on('error', (err) => { throw new Error(err.toString()); });
    page.on('pageerror', (err) => { throw new Error(err.toString()); });
    page.on('console', (msg) => {
      if (options.haltOnConsoleError && msg.type() === "error") {
        throw new Error(JSON.stringify(msg));
      }
    });

    await page.setViewport({
      width: Number(options.width),
      height: Number(options.height)
    });
    await page.setContent(html, { waitUntil: 'networkidle0', timeout: Number(options.timeout) });

    let imageBinary = await page.screenshot({
      type: options.type || 'png',
      quality: options.type === 'jpeg' ? Number(options.quality || 100) : undefined,
      fullPage: true
    });

    if (options.imagemin) {
      const imagemin = require("imagemin");
      const imageminJpegtran = require("imagemin-jpegtran");
      const imageminOptipng = require("imagemin-optipng");
      const plugins = []

      if (options.type === 'jpeg') {
        plugins.push(imageminJpegtran());
      } else {
        plugins.push(imageminOptipng());
      }

      return imagemin.buffer(imageBinary, { plugins });
    } else {
      return imageBinary;
    }
  } finally {
    await browser.close();
  }
}

module.exports = { osmsm };
